File: labs-opensource/M08-conversation-management/starter/conversation_manager.js
/**
 * M08 Lab - Steps 1+2: ConversationManager + SlidingWindowManager
 * ================================================================
 * Run: node conversation_manager.js
 */

import OpenAI from "openai";

import { pathToFileURL } from "node:url";
const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });

// ── Step 1: Basic ConversationManager ────────────────────────
export class ConversationManager {
  constructor({ systemPrompt = "You are a helpful assistant.", model = "mistral" } = {}) {
    this.systemPrompt = systemPrompt;
    this.model = model;
    this.messages = [];
    this.totalInputTokens = 0;
    this.totalOutputTokens = 0;
  }

  addUserMessage(content) {
    this.messages.push({ role: "user", content });
  }

  addAssistantMessage(content) {
    this.messages.push({ role: "assistant", content });
  }

  /** Return messages with system prompt prepended. (COMPLETE) */
  getMessages() {
    return [{ role: "system", content: this.systemPrompt }, ...this.messages];
  }

  /**
   * Send a message and get the model's response.
   *
   * TODO:
   * 1. this.addUserMessage(userMessage);
   * 2. response = await client.chat.completions.create({ model: this.model,
   *      messages: this.getMessages() });
   * 3. If response.usage: accumulate prompt_tokens / completion_tokens
   *    into this.totalInputTokens / this.totalOutputTokens
   * 4. Push + return the assistant text
   * 5. On error: this.messages.pop() (remove the failed user message),
   *    then throw new Error(`API call failed: ${error.message}`)
   */
  async send(userMessage) {
    // TODO: implement
  }

  getTokenUsage() {
    return {
      totalInput: this.totalInputTokens,
      totalOutput: this.totalOutputTokens,
      messages: this.messages.length,
    };
  }
}

// ── Step 2: SlidingWindowManager ─────────────────────────────
export class SlidingWindowManager extends ConversationManager {
  constructor({ windowSize = 10, ...opts } = {}) {
    super(opts);
    this.windowSize = windowSize;
  }

  /**
   * Return system prompt + only the most recent N messages.
   *
   * TODO:
   * 1. let windowed = the last this.windowSize entries of this.messages
   *    (or all of them if there are fewer)
   * 2. GOTCHA: if windowed starts with an "assistant" message, drop it —
   *    history sent to the API should start with a user turn
   * 3. Return [{ role: "system", ... }, ...windowed]
   */
  getMessages() {
    // TODO: implement
  }
}

// ── Test harness (COMPLETE) ──────────────────────────────────
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  console.log("=".repeat(50));
  console.log("TEST 1: Basic ConversationManager (full history)");
  console.log("=".repeat(50));
  const mgr = new ConversationManager({ systemPrompt: "You are a concise coding tutor. Answer in 1-2 sentences." });
  for (const q of ["What is a list comprehension in Python?", "Show me an example with filtering."]) {
    console.log(`\nQ: ${q}`);
    console.log(`A: ${(await mgr.send(q)).slice(0, 150)}`);
    console.log(`   usage so far:`, mgr.getTokenUsage());
  }

  console.log("\n" + "=".repeat(50));
  console.log("TEST 2: SlidingWindowManager (windowSize=6)");
  console.log("=".repeat(50));
  const win = new SlidingWindowManager({ windowSize: 6, systemPrompt: "You are a concise coding tutor." });
  for (const q of ["What is Python?", "What are variables?", "Explain loops.",
                   "What are functions?", "Explain classes.", "What is inheritance?"]) {
    await win.send(q);
    console.log(`Q: ${q}  (stored=${win.messages.length}, sent=${win.getMessages().length - 1})`);
  }

  console.log(`\nStored: ${win.messages.length} messages`);
  console.log(`Sent (excl. system): ${win.getMessages().length - 1} messages`);
}

File: labs-opensource/M08-conversation-management/starter/conversation_manager.py
"""
M08 Lab - Steps 1+2: ConversationManager + SlidingWindowManager
================================================================
Run: python conversation_manager.py
"""

from dataclasses import dataclass, field

from openai import OpenAI

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")


# ── Step 1: Basic ConversationManager ────────────────────────
@dataclass
class ConversationManager:
    """Full-history conversation manager with token tracking."""

    system_prompt: str = "You are a helpful assistant."
    model: str = "mistral"
    messages: list = field(default_factory=list)

    def __post_init__(self):
        self.total_input_tokens = 0
        self.total_output_tokens = 0

    def add_user_message(self, content: str) -> None:
        self.messages.append({"role": "user", "content": content})

    def add_assistant_message(self, content: str) -> None:
        self.messages.append({"role": "assistant", "content": content})

    def get_messages(self) -> list:
        """Return messages with system prompt prepended. (COMPLETE)"""
        return [{"role": "system", "content": self.system_prompt}] + list(self.messages)

    def send(self, user_message: str) -> str:
        """Send a message and get the model's response.

        TODO:
        1. self.add_user_message(user_message)
        2. response = client.chat.completions.create(model=self.model,
               messages=self.get_messages())
        3. If response.usage: accumulate prompt_tokens / completion_tokens
           into self.total_input_tokens / self.total_output_tokens
        4. Append + return the assistant text
        5. On exception: self.messages.pop() (remove the failed user message),
           then raise RuntimeError(f"API call failed: {e}")
        """
        pass  # Remove this line when you add your code

    def get_token_usage(self) -> dict:
        return {
            "total_input": self.total_input_tokens,
            "total_output": self.total_output_tokens,
            "messages": len(self.messages),
        }


# ── Step 2: SlidingWindowManager ─────────────────────────────
class SlidingWindowManager(ConversationManager):
    """Stores everything, but only SENDS the most recent N messages."""

    def __init__(self, window_size: int = 10, **kwargs):
        super().__init__(**kwargs)
        self.window_size = window_size

    def get_messages(self) -> list:
        """Return system prompt + only the most recent N messages.

        TODO:
        1. windowed = the last self.window_size entries of self.messages
           (or all of them if there are fewer)
        2. GOTCHA: if windowed starts with an "assistant" message, drop it —
           history sent to the API should start with a user turn
        3. Return [{"role": "system", ...}] + windowed
        """
        pass  # Remove this line when you add your code


# ── Test harness (COMPLETE) ──────────────────────────────────
if __name__ == "__main__":
    print("=" * 50)
    print("TEST 1: Basic ConversationManager (full history)")
    print("=" * 50)
    mgr = ConversationManager(system_prompt="You are a concise coding tutor. Answer in 1-2 sentences.")
    for q in ["What is a list comprehension in Python?", "Show me an example with filtering."]:
        print(f"\nQ: {q}")
        print(f"A: {mgr.send(q)[:150]}")
        print(f"   usage so far: {mgr.get_token_usage()}")

    print("\n" + "=" * 50)
    print("TEST 2: SlidingWindowManager (window_size=6)")
    print("=" * 50)
    win = SlidingWindowManager(window_size=6, system_prompt="You are a concise coding tutor.")
    for q in ["What is Python?", "What are variables?", "Explain loops.",
              "What are functions?", "Explain classes.", "What is inheritance?"]:
        win.send(q)
        print(f"Q: {q}  (stored={len(win.messages)}, sent={len(win.get_messages()) - 1})")

    print(f"\nStored: {len(win.messages)} messages")
    print(f"Sent (excl. system): {len(win.get_messages()) - 1} messages")
